### PR TITLE
[Snippets][CPU] Added dynamic emitters and nodes for control-flow

### PR DESCRIPTION
**Pull request by a-sidorova**
### Details:
 - *Split control-flow nodes (`Kernel`, `LoopBegin`, `LoopEnd`) to static and dynamic.*
 - *Split control-flow emitters to static and dynamic*

### Tickets:
 - *113212*

### Prereuisites:
- https://github.com/openvinotoolkit/openvino/pull/21934

### CODE REVIEW

File: src/common/snippets/include/snippets/lowered/linear_ir.hpp [hunk 1]
```diff
@@ -139,6 +139,9 @@ class LinearIR {
     const std::shared_ptr<ShapeInferSnippetsNode>& get_shape_infer_instance() const {return m_shape_infer; }
     VectorDims get_master_shape() const;
 
+    // TODO: Need to add smart algorithm
+    bool is_dynamic() const { return false; }
```
Comment (a-sidorova) [inline src/common/snippets/include/snippets/lowered/linear_ir.hpp:143]: Will be implemented in the next PRs

File: src/common/snippets/src/lowered/pass/insert_loops.cpp [hunk 2]
```diff
@@ -88,26 +77,45 @@ void InsertLoops::insertion(LinearIR& linear_ir, const LinearIR::LoopManagerPtr&
     init_params(loop_entries);
     init_params(loop_exits);
 
-    const auto& loop_begin = std::make_shared<op::LoopBegin>();
+    // Should be inited by LoopInfo
+    const auto is_dynamic_loop = false;
```
Comment (a-sidorova) [inline src/common/snippets/src/lowered/pass/insert_loops.cpp:81]: Will be implemented in the next PRs

File: src/plugins/intel_cpu/src/emitters/snippets/x64/jit_loop_emitters.cpp [hunk 3]
```diff
@@ -12,110 +13,238 @@ using namespace dnnl::impl::cpu::x64;
...
+std::shared_ptr<ov::snippets::op::LoopEnd> jit_loop_begin_emitter::get_loop_end(const ov::snippets::lowered::ExpressionPtr& expr) {
     OPENVINO_ASSERT(expr->get_output_port_connectors().size() == 1, "jit_loop_begin_emitter has invalid LoopBegin expression configuration");
     const auto& consumers = expr->get_output_port_connector(0)->get_consumers();
     OPENVINO_ASSERT(consumers.size() == 1, "jit_loop_begin_emitter has invalid LoopBegin expression configuration");
-    const auto loop_end = ov::as_type_ptr<snippets::op::LoopEnd>(consumers.cbegin()->get_expr()->get_node());
-    OPENVINO_ASSERT(loop_end != nullptr, "jit_loop_begin_emitter has invalid LoopBegin expression configuration");
-    return loop_end;
+    const auto loop_end_expr = consumers.cbegin()->get_expr();
+    OPENVINO_ASSERT(ov::is_type<snippets::op::LoopEnd>(loop_end_expr->get_node()), "jit_loop_begin_emitter has invalid LoopBegin expression configuration");
+    return ov::as_type_ptr<snippets::op::LoopEnd>(loop_end_expr->get_node());
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/emitters/snippets/x64/jit_loop_emitters.cpp:35]: It this implementation better than the old one? It looks like the new one does exactly the same as the old one, but performs both `is_type` and `as_type_ptr` casts (which is clearly redundant). What is the reason behind this change?

File: src/common/snippets/src/generator.cpp [hunk 4]
```diff
@@ -42,10 +44,21 @@ void Generator::generate(lowered::LinearIR& linear_ir, LoweringResult& result, c
...
-    std::shared_ptr<Emitter> kernel = target->get(op::Kernel::get_type_info_static())(loops2DKernelExpr);
+
+    const auto contains_dynamic_exprs = std::any_of(linear_ir.cbegin(), linear_ir.cend(),
+                       [](const lowered::ExpressionPtr& expr) {
+                            return ov::is_type<op::LoopBeginDynamic>(expr->get_node()) ||
+                                   ov::is_type<op::LoopEndDynamic>(expr->get_node());
+                       });
+
+    std::shared_ptr<op::Kernel> kernel_op = nullptr;
+    if (linear_ir.is_dynamic() || contains_dynamic_exprs)
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/generator.cpp]: As I understand we cannot make helper in `jit_kenrel_emitter.hpp` since it's CPU related code.
Anyway I agree that it'd be great if we have some builder for kernel. I added it to kernel node impl. Can you take a look please?

As for two conditions, I tried to make generic check (if we have internal dynamism). But seems like it's redundant at the moment. So let's leave only `contains_dynamic_exprs`

File: src/common/snippets/src/lowered/pass/cleanup_loop_offsets.cpp [hunk 5]
```diff
@@ -22,54 +22,60 @@ bool CleanupLoopOffsets::run(LinearIR& linear_ir) {
...
-                    is_modified = true;
-                }
-                if (auto outer_loop_end = as_type_ptr<op::LoopEnd>(next_node)) {
-                    auto fin_offsets = loop_end->get_finalization_offsets();
-                    std::unordered_map<PortConnectorPtr, size_t> per_port_connector_offset;
-                    const auto& loop_inputs = expr_it->get()->get_input_port_connectors();
-                    for (size_t i = 0; i < fin_offsets.size(); i++)
-                        per_port_connector_offset[loop_inputs[i]] = i;
+        if (auto loop_end = as_type_ptr<op::LoopEndStatic>(node)) {
+            auto next_expr_it = std::next(expr_it);
```
Comment (a-sidorova) [inline src/common/snippets/src/lowered/pass/cleanup_loop_offsets.cpp:26]: Removed redundant tab. There are only two changes: added `if (!outer_is_incremented[i])` and introduced `LoopEndStatic`

**Comment by a-sidorova** [comment]
@IvanNovoselov could you take a look one more time please? Thank you!

**Comment by a-sidorova** [comment]
@IvanNovoselov kind reminder :blush:

File: src/plugins/intel_cpu/src/emitters/snippets/x64/jit_kernel_emitter.hpp [hunk 6]
```diff
@@ -47,39 +85,79 @@ struct jit_snippets_compile_args {
 class jit_kernel_emitter : public jit_container_emitter {
 public:
     jit_kernel_emitter(dnnl::impl::cpu::x64::jit_generator* h, dnnl::impl::cpu::x64::cpu_isa_t isa,
-                       const ov::snippets::lowered::ExpressionPtr& expr);
+                       const ov::snippets::lowered::ExpressionPtr& expr, size_t reg_runtime_params_idx);
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/emitters/snippets/x64/jit_kernel_emitter.hpp:88]: Are you sure we need this complication?
Would it be easier to change the orger of runtime arguments for static kernels:
` callable(&call_args, indexes)`  instead of ` callable(indexes, &call_args)` 
while dynamic impl would be called ` callable(&call_args)`

It's better to align static and dynamic kernels if we can. What do you think?

Comment (a-sidorova) [reply src/plugins/intel_cpu/src/emitters/snippets/x64/jit_kernel_emitter.hpp]: OMG, it's obvious, simpler and better! I don't know why I didn't thank about that 😆 
Thanks a lot!

File: src/common/snippets/src/op/kernel.cpp [hunk 7]
```diff
@@ -4,12 +4,38 @@
...
 
+#include "snippets/op/loop.hpp"
+
 namespace ov {
 namespace snippets {
 namespace op {
 
 Kernel::Kernel(lowered::LinearIR nested) : Op(), region(std::move(nested)) {}
 
+std::shared_ptr<Kernel> Kernel::make_kernel(const lowered::LinearIR& region) {
```
[the remark this answers is not in the document]
Comment (a-sidorova) [reply src/common/snippets/src/op/kernel.cpp:15]: Discussed offline: added class field `LinearIR::m_is_dynamic` that is inited in `LinearIR` construction. Thanks!

**Comment by a-sidorova** [comment]
@IvanNovoselov thank you for the review! I've applied the comments so could you take a look again please?

**Review by IvanNovoselov — APPROVED** [review]
👍

File: src/common/snippets/src/op/loop.cpp [hunk 8]
```diff
@@ -121,71 +88,127 @@ size_t LoopEnd::get_id() const {
...
     if (m_finalization_offsets.empty())
         m_finalization_offsets.resize(io_size, 0);
-    set_output_type(0, element::f32, ov::PartialShape{ov::Shape{}});
 }
 
-bool LoopEnd::visit_attributes(AttributeVisitor &visitor) {
-    visitor.on_attribute("work_amount", m_work_amount);
-    visitor.on_attribute("increment", m_work_amount_increment);
+bool LoopEndStatic::visit_attributes(AttributeVisitor &visitor) {
     visitor.on_attribute("ptr_incr", m_ptr_increments);
```
Comment (v-Golubev) [inline src/common/snippets/src/op/loop.cpp:155]: Should we visit `m_work_amount` too?

**Review by v-Golubev — APPROVED** [review]
LGTM

File: src/common/snippets/include/snippets/op/loop.hpp [hunk 9]
```diff
@@ -65,51 +76,79 @@ class LoopBegin : public LoopBase {
...
-    size_t get_work_amount() const;
-    size_t get_increment() const;
-    bool get_evaluate_once() const;
-    size_t get_id() const;
-    bool visit_attributes(AttributeVisitor& visitor) override;
+    void set_id(size_t id);
 
-private:
+protected:
     std::vector<bool> m_is_incremented = {};
```
[the remark this answers is not in the document]
Comment (v-Golubev) [reply src/common/snippets/include/snippets/op/loop.hpp:101]: Thanks for the explanation!

**Review by dmitry-gorokhov — APPROVED** [review]
